builtin/logical/database/credentials.go

```diff
@@ -170,3 +175,187 @@ func (kg rsaKeyGenerator) configMap() (map[string]interface{}, error) {
+}
+
+// newClientCertificateGenerator returns a new ClientCertificateGenerator
+// using the given config. Default values will be set on the returned
+// ClientCertificateGenerator if not provided in the config.
+func newClientCertificateGenerator(config map[string]interface{}) (ClientCertificateGenerator, error) {
+	var cg ClientCertificateGenerator
+	if err := mapstructure.WeakDecode(config, &cg); err != nil {
+		return cg, err
+	}
+
+	switch cg.KeyType {
+	case "":
+		cg.KeyType = "ec"
+	case "rsa":
+		switch cg.KeyBits {
+		case 0:
+			cg.KeyBits = 2048
+		case 2048, 3072, 4096:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ec":
+		switch cg.KeyBits {
+		case 0:
+			cg.KeyBits = 256
+		case 224, 256, 384, 521:
+		default:
+			return cg, fmt.Errorf("invalid key_bits")
+		}
+	case "ed25519":
+	// key_bits ignored
+	default:
+		return cg, fmt.Errorf("invalid key_type")
+	}
+
+	switch cg.SignatureBits {
+	case 0, 256, 384, 512:
+	default:
+		return cg, fmt.Errorf("invalid signature_bits")
+	}
+
+	if cg.CommonNameTemplate == "" {
+		return cg, fmt.Errorf("missing required common_name_template")
+	}
+
+	// Validate the common name template
+	t, err := template.NewTemplate(template.Template(cg.CommonNameTemplate))
+	if err != nil {
+		return cg, fmt.Errorf("failed to create template: %w", err)
+	}
+
+	_, err = t.Generate(dbplugin.UsernameMetadata{})
+	if err != nil {
+		return cg, fmt.Errorf("invalid common_name_template: %w", err)
+	}
+	cg.cnProducer = t
+
+	if cg.CACert == "" {
+		return cg, fmt.Errorf("missing required ca_cert")
+	}
+	if cg.CAPrivateKey == "" {
+		return cg, fmt.Errorf("missing required ca_private_key")
+	}
+	parsedBundle, err := certutil.ParsePEMBundle(strings.Join([]string{cg.CACert, cg.CAPrivateKey}, "\n"))
+	if err != nil {
+		return cg, err
+	}
+	if parsedBundle.PrivateKey == nil {
+		return cg, fmt.Errorf("private key not found in the PEM bundle")
+	}
+	if parsedBundle.PrivateKeyType == certutil.UnknownPrivateKey {
+		return cg, fmt.Errorf("unknown private key found in the PEM bundle")
+	}
+	if parsedBundle.Certificate == nil {
+		return cg, fmt.Errorf("certificate not found in the PEM bundle")
+	}
+	if !parsedBundle.Certificate.IsCA {
+		return cg, fmt.Errorf("the given certificate is not marked for CA use")
+	}
+	if !parsedBundle.Certificate.BasicConstraintsValid {
+		return cg, fmt.Errorf("the given certificate does not meet basic constraints for CA use")
+	}
+
+	certBundle, err := parsedBundle.ToCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("error converting raw values into cert bundle: %w", err)
+	}
+
+	parsedCABundle, err := certBundle.ToParsedCertBundle()
+	if err != nil {
+		return cg, fmt.Errorf("failed to parse cert bundle: %w", err)
+	}
+	cg.parsedCABundle = parsedCABundle
+
+	return cg, nil
+}
+
+func (cg *ClientCertificateGenerator) generate(r io.Reader, expiration time.Time, userMeta dbplugin.UsernameMetadata) (*certutil.CertBundle, string, error) {
```

Comment:
nit: is it possible to write a test for this? Understand if it is hard to do with the SDK 🙏🏼

Reply:
I'll create a separate task for this, it will require some more time